### PR TITLE
feat(android-sdk): social native sign in

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

feat(android-sdk): social sign in
This PR includes:
* Social sign-in process of web and native(alipay & wechat)
* Implement the communication conventions between the UI and native (LOG-2125)
* Use `consent` as the `prompt` value and clear the WebView cookies on signing out.
* Support configurations for the native social sign-in process.

## Linear Issue Reference

* LOG-2083
* LOG-2168
* LOG-2185
* LOG-2124
* LOG-2183

#### Alipay
https://user-images.githubusercontent.com/10806653/162177009-f977c5fd-67f2-453e-9a4f-03cbb482ee62.mp4

#### Wechat
https://user-images.githubusercontent.com/10806653/162176576-1d23403d-e94e-4e83-ba46-7c86e06203dc.mp4

#### Alipay
alipay
### Wechat
wechat

### Inject the social config to the UI.
image

### CODE REVIEW

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    fun postSocialMessage(socialType: String) {
+        // TODO - Conventions Between UI and Native
+        webView.post {
+            handleSocial(socialType)
+        }
+    }
+
+    private fun handleSocial(socialType: String) {
+        println("Social == Social Sign In Click")
+        println("Social == SocialType: $socialType")
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:28]: should remove?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import io.logto.sdk.android.auth.social.alipay.AlipaySocialSession
+import io.logto.sdk.android.auth.social.web.WebSocialSession
+import io.logto.sdk.android.auth.social.wechat.WechatSocialSession
+
+class LogtoWebViewSocialHandler(
+    private val webView: WebView,
+    private val hostActivity: Activity,
+) {
+    companion object {
+        const val SOCIAL_HANDLER_NAME = "SocialHandler"
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:15]: just `NAME` is cool?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        // TODO - Conventions Between UI and Native
+        webView.post {
+            handleSocial(socialType)
+        }
+    }
+
+    private fun handleSocial(socialType: String) {
+        println("Social == Social Sign In Click")
+        println("Social == SocialType: $socialType")
+        // TODO - Create Social Sessions Dynamically
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:29]: what's this exactly? should we refer to a linear issue to provide more context?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package io.logto.sdk.android.auth.social.web
+
+import android.app.Activity
+import android.content.Intent
+import android.net.Uri
+import io.logto.sdk.android.completion.Completion
+
+class WebSocialSession(
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:8]: should we define a protocol for social sessions like Swift SDK?

**Review by gao-sun — COMMENTED** [review]
would be great if you can brief the workflow to engineers tomorrow

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt (hunk 3 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:29]: LOG-2124

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt (hunk 4 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:8]: LOG-2125

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt (hunk 1 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt]: updated

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt (hunk 2 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt]: updated

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt (hunk 3 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt]: done.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt (hunk 4 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:8]: done.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    // TODO - LOG-2178: Add `isAvailable` for social plugins
+    // Use ["wechat-native", "alipay"] temporary
+    fun getInjectSocialScript() = """
+        window.logtoNativeSdk = {
+            platform: 'android',
+            getPostMessage: () => (data) => {
+                window.$NAME.postMessage(JSON.stringify(data));
+            },
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:31]:
```suggestion
            getPostMessage: () => window.$NAME.postMessage,
```

File: android-sdk/android/src/main/res/values/logto.xml [hunk 6]
```diff
@@ -1,4 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-    <string name="logto_redirect_scheme">logto_redirect_scheme_placeholder</string>
+    <string name="logto_redirect_scheme" />
+    <string name="wechat_app_id" />
+    <string name="alipay_app_id" />
```
Comment (gao-sun) [inline android-sdk/android/src/main/res/values/logto.xml:5]: whats this?

Comment (gao-sun) [reply android-sdk/android/src/main/res/values/logto.xml:5]: this should read from webview postmessage instead of defining here.

File: android-social/alipay/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+group = "io.logto.sdk"
+version = libs.versions.logtoSdk.get()
+
+configurations.create("default")
+artifacts.add("default", file("alipaySdk-15.7.9-20200727142846.aar"))
```
Comment (gao-sun) [inline android-social/alipay/build.gradle.kts:5]: we should not put it in source code per discussion in swift sdk

File: kotlin-sdk/kotlin/src/main/kotlin/io/logto/sdk/core/constant/PromptValue.kt [hunk 8]
```diff
@@ -1,5 +1,5 @@
 package io.logto.sdk.core.constant
 
 object PromptValue {
-    const val LOGIN = "login"
+    const val CONSENT = "consent"
```
Comment (gao-sun) [inline kotlin-sdk/kotlin/src/main/kotlin/io/logto/sdk/core/constant/PromptValue.kt:4]: should be a separate PR, let's notice next time

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                    completion.onComplete(null, (it as SendAuth.Resp).code)
+                    return
+                }
+            }
+        }
+        completion.onComplete(LogtoException(LogtoException.Message.WECHAT_AUTH_FAILED), null)
+    }
+
+    fun handleMissingAppIdError() {
+        completion.onComplete(LogtoException(LogtoException.Message.WECHAT_APP_ID_NO_FOUND), null)
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt:44]: this can be inline

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialHelper.kt [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+import android.app.Activity
+import com.tencent.mm.opensdk.openapi.IWXAPI
+import com.tencent.mm.opensdk.openapi.WXAPIFactory
+
+internal object WechatSocialHelper {
+    fun getApi(context: Activity, appId: String): IWXAPI =
+        WXAPIFactory.createWXAPI(context, appId).apply {
+            registerApp(appId)
+        }
+}
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialHelper.kt:12]: don't need this extra object? we can just put it into result activity.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt (hunk 4 above)
Comment (gao-sun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:8]: where's the protocol?

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: android-social/alipay/build.gradle.kts (hunk 7 above)
Comment (gao-sun) [reply android-social/alipay/build.gradle.kts:5]: you can leverage https://github.com/logto-io/social-sdks if needed

File: android-sdk/android/src/main/res/values/logto.xml (hunk 6 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/res/values/logto.xml:5]: These are placeholders for our sdk, we need these when implementing our sdk.
This config in the application will override these placeholders.
We don't have a convention about the way we use the app id?

File: android-social/alipay/build.gradle.kts (hunk 7 above)
Comment (xiaoyijun) [reply android-social/alipay/build.gradle.kts:5]: We need to publish it if we put this into another repo, it's very inconvenient to share the libs between multi repos in the development process. This is a very common practice in Gradle.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialHelper.kt (hunk 10 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialHelper.kt]: updated

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt (hunk 9 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt:47]: updated

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt (hunk 4 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:9]: I misunderstand the `protocol` you mentioned. I thought we were talking about the convention about the UI & the native side.
Updated, PTAL.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import io.logto.sdk.core.util.GenerateUtils
+
+class WechatSocialSession(
+    override val context: Activity,
+    override val redirectTo: String,
+    override val callbackUri: String,
+    override val completion: Completion<String>,
+) : SocialSession {
+    companion object {
+        const val APP_ID = "app_id"
```
Comment (simeng-li) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt:21]: could this be defined in the SocialSession interface? looks like the same for all the social sessions?

**Comment by simeng-li** [comment]
wondering is there a way we could add some tests for those codes?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt (hunk 4 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:9]: We have a `SocialSession` interface. this `WebSocialSession` implements the interface.

File: android-sdk/android/src/main/AndroidManifest.xml [hunk 12]
```diff
@@ -5,26 +5,41 @@
...
+
+        <activity android:name="com.alipay.sdk.app.AlipayResultActivity"
+            android:exported="true">
+            <intent-filter>
+                <action android:name="android.intent.action.VIEW"/>
+                <category android:name="android.intent.category.DEFAULT"/>
+                <category android:name="android.intent.category.BROWSABLE"/>
+                <data android:scheme="${applicationId}.logto-alipay-callback"/>
             </intent-filter>
         </activity>
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/AndroidManifest.xml:44]: Please refer to [the Slack channel](https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1649929272931289).

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            const val CALLBACK_URI = "callbackUri"
+        }
+    }
+
+    // TODO - LOG-2178: Add `isAvailable` for social plugins
+    // Use ["wechat-native", "alipay"] temporary
+    fun getInjectSocialScript() = """
+        window.logtoNativeSdk = {
+            platform: 'android',
+            getPostMessage: () => (data) => window.$NAME.postMessage(JSON.stringify(data)),
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:28]: Refer to the comment above: https://github.com/logto-io/kotlin/pull/132#discussion_r850492856

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            supportedSocialConnectorIds: ["wechat-native", "alipay"],
+            callbackUriScheme: '${hostActivity.packageName}.logto-web-social-callback',
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:30]: Per discussion offline last time, Android cannot hard code this, so we have a convention about the data passing to the UI page, which includes the `callbackUriScheme` field.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+            // TODO - LOG-2186: Handle Errors in Social Sign in Process
+            return
+        }
+
+        webView.post {
+            val socialSession = SocialSessionFactory.createSocialSession(
+                scheme = scheme,
+                context = hostActivity,
+                redirectTo = redirectTo,
+                callbackUri = callbackUri,
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:52]: Align with the Swift Sdk.
Maybe we need a discussion on this.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    override val callbackUri: String,
+    override val completion: Completion<String>,
+) : SocialSession {
+    override fun start() {
+        WebSocialResultActivity.registerSession(this)
+        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(redirectTo)))
+    }
+
+    fun handleResult(data: Uri) {
+        completion.onComplete(null, data.toString())
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/web/WebSocialSession.kt:21]: In our case, the registered redirect URI will be the signing page of the UI. May be we don't need to handle that in the native side?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialResultActivity.kt [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val redirectTo = wechatSocialSession[WECHAT_SOCIAL_SESSION_KEY]?.redirectTo
+        val appId = if (redirectTo == null) {
+            // Note: Uri.parse will throw NullPointerException if redirectTo is null
+            null
+        } else {
+            Uri.parse(redirectTo).getQueryParameter(WechatSocialSession.APP_ID)
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialResultActivity.kt:28]: No. It's commonly used in kotlin, for not like other program languages, the `if-else` is an expression in kotlin.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt (hunk 11 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt:21]: Shall we define a static value inside an interface?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/alipay/AlipaySocialSession.kt [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+import com.alipay.sdk.app.OpenAuthTask
+import io.logto.sdk.android.auth.social.SocialSession
+import io.logto.sdk.android.completion.Completion
+import io.logto.sdk.android.exception.LogtoException
+import io.logto.sdk.core.util.GenerateUtils
+
+class AlipaySocialSession(
+    override val context: Activity,
+    override val redirectTo: String,
+    override val callbackUri: String,
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/alipay/AlipaySocialSession.kt:14]: Ditto

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt (hunk 11 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt:21]: I think it's good to keep it separated, for it may be "client_id" in some cases.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt (hunk 5 above)
Comment (gao-sun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt]: the suggestion is not passing js object either..

File: android-social/alipay/build.gradle.kts [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+
+val alipaySdkName = "alipaySdk-15.7.9-20200727142846.aar"
+val sdk = file(alipaySdkName)
+if (!sdk.exists()) {
+    val sdkUri = "https://github.com/logto-io/social-sdks/raw/8aa90c6eaf68d1c4f2278bf7cd339b7a37c4c2c0/alipay/android/alipaySdk-15.7.9-20200727142846.aar"
+    uri(sdkUri).toURL().openStream().use {
+        it.copyTo(java.io.FileOutputStream(sdk))
+    }
+}
+
```
Comment (gao-sun) [inline android-social/alipay/build.gradle.kts:12]: this can be refactored to something like:

``` kt
val sdks = [["alipay-sdk.arr", "https://some-url"]]

// then loop `sdks` to download every sdk to `arr/` or some centralized folder

Comment (gao-sun) [reply android-social/alipay/build.gradle.kts:12]: can add an issue and do it later though

File: android-sdk/android/src/main/res/values/logto.xml [hunk 20]
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-    <string name="logto_redirect_scheme">logto_redirect_scheme_placeholder</string>
+    <string name="logto_redirect_scheme" />
```
Comment (gao-sun) [inline android-sdk/android/src/main/res/values/logto.xml:3]: do we still need this if we can have a convention like `logto-callback://:appId/:connectorId`?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/alipay/AlipaySocialSession.kt [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    override val completion: Completion<String>,
+) : SocialSession {
+    companion object {
+        private const val APP_ID = "app_id"
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/alipay/AlipaySocialSession.kt:19]: i don't think we need to define this since:
1. it varies by social identity provider
2. we only use the key once

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/SocialSession.kt [hunk 22]
```diff
@@ -0,0 +1,13 @@
+package io.logto.sdk.android.auth.social
+
+import android.app.Activity
+import io.logto.sdk.android.completion.Completion
+
+interface SocialSession {
+    val context: Activity
+    val redirectTo: String
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/SocialSession.kt:8]: discussion: should we create session with `redirectTo` and `callbackUri` or like swift SDK, pass them in `.start()`

cc @logto-io/eng

File: android-social/alipay/build.gradle.kts (hunk 19 above)
Comment (xiaoyijun) [reply android-social/alipay/build.gradle.kts]: updated

File: android-social/alipay/build.gradle.kts (hunk 7 above)
Comment (xiaoyijun) [reply android-social/alipay/build.gradle.kts]: updated

File: android-social/alipay/build.gradle.kts (hunk 19 above)
Comment (xiaoyijun) [reply android-social/alipay/build.gradle.kts]: updated

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/alipay/AlipaySocialSession.kt (hunk 21 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/alipay/AlipaySocialSession.kt]: updated

File: android-sample-kotlin/app/src/main/AndroidManifest.xml [hunk 23]
```diff
@@ -1,9 +1,13 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="io.logto.demo">
+    package="io.logto.android.sample4k">
```
Comment (demonzoo) [inline android-sample-kotlin/app/src/main/AndroidManifest.xml:3]: What does `sample4k` stand for, "sample for kotlin"?
How about just "io.log.kotlin.sample"?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialResultActivity.kt [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+import android.os.Bundle
+import com.tencent.mm.opensdk.modelbase.BaseReq
+import com.tencent.mm.opensdk.modelbase.BaseResp
+import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler
+import com.tencent.mm.opensdk.openapi.WXAPIFactory
+
+open class WechatSocialResultActivity : Activity(), IWXAPIEventHandler {
+    companion object {
+        private const val WECHAT_SOCIAL_SESSION_KEY = "wechat"
+        private val wechatSocialSession = mutableMapOf<String, WechatSocialSession>()
```
Comment (IceHe) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialResultActivity.kt:14]: There seems to be only one key-value pair in this map.
Why not just use a `wechatSocialSession` for now?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+        logtoAuthSession.remove(uri.scheme)?.handleCallbackUri(uri)
+    }
+
+    fun handleUserCancel() {
+        logtoAuthSession.remove(logtoAuthCallbackUriScheme)?.handleUserCancel()
+    }
+
+    fun isLogtoAuthCallbackUriScheme(uriScheme: String?): Boolean {
+        return uriScheme?.let {
+            logtoAuthSession.containsKey(it)
```
Comment (demonzoo) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt:24]: 'contains` does not sound strict to me, is there a `startsWith` method?

File: android-sdk/android/src/main/AndroidManifest.xml [hunk 26]
```diff
@@ -5,26 +5,31 @@
...
+            android:exported="true">
             <intent-filter>
-                <action android:name="android.intent.action.VIEW" />
-                <category android:name="android.intent.category.DEFAULT" />
-                <category android:name="android.intent.category.BROWSABLE" />
-                <data android:scheme="logto-callback" />
+                <action android:name="android.intent.action.VIEW"/>
+                <category android:name="android.intent.category.DEFAULT"/>
+                <category android:name="android.intent.category.BROWSABLE"/>
+                <data android:scheme="${applicationId}.logto-callback-alipay"/>
```
Comment (IceHe) [inline android-sdk/android/src/main/AndroidManifest.xml:32]: The code style seems not to be consistent.
Better separate a PR to format code style.

See the code above
image

**Review by IceHe — CHANGES_REQUESTED** [review]

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewAuthClient.kt [hunk 27]
```diff
@@ -1,28 +1,28 @@
...
-import io.logto.sdk.android.auth.LogtoAuthManager
 
-class WebViewAuthClient(
+class LogtoWebViewAuthClient(
     private val hostActivity: Activity,
+    private val injectScript: String,
 ) : WebViewClient() {
+    override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
+        super.onPageStarted(view, url, favicon)
+        view?.evaluateJavascript(injectScript, null)
```
Comment (demonzoo) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewAuthClient.kt:15]: What JavaScript code do we inject into the webview?

Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewAuthClient.kt:15]: https://github.com/logto-io/kotlin/pull/132/files#diff-be0402f1c7797833d11f1933f48fdcb18b2c19f8a6fb94b05664f8d2b92d1c14R26-R30

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt (hunk 25 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt:24]: `logtoAuthSession` is a map, not a string.

File: android-sdk/android/src/main/AndroidManifest.xml [hunk 28]
```diff
@@ -5,26 +5,31 @@
...
-        <activity android:name=".auth.CallbackUriActivity"
-                  android:exported="true">
+        <activity android:name=".auth.social.web.WebSocialResultActivity"
+            android:exported="true">
             <intent-filter>
                 <action android:name="android.intent.action.VIEW" />
                 <category android:name="android.intent.category.DEFAULT" />
                 <category android:name="android.intent.category.BROWSABLE" />
-                <data android:scheme="@string/logto_redirect_scheme" />
+                <data android:scheme="${applicationId}.logto-callback-web" />
```
Comment (gao-sun) [inline android-sdk/android/src/main/AndroidManifest.xml:22]: will it be better to use fixed scheme and leverage host/path say `logto-callback://${appId}/web`

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt [hunk 29]
```diff
@@ -0,0 +1,36 @@
+package io.logto.sdk.android.auth.logto
+
+import android.annotation.SuppressLint
+import android.net.Uri
+
+internal object LogtoAuthManager {
+    private var logtoAuthCallbackUriScheme: String? = null
+    @SuppressLint("StaticFieldLeak")
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt:8]: what's this? should we add some comments for it?

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt [hunk 30]
```diff
@@ -0,0 +1,36 @@
+package io.logto.sdk.android.auth.logto
+
+import android.annotation.SuppressLint
+import android.net.Uri
+
+internal object LogtoAuthManager {
+    private var logtoAuthCallbackUriScheme: String? = null
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt:7]: if we just use fixed scheme then there's no need to define a var here

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+            const val CALLBACK_URI = "callbackUri"
+        }
+    }
+
+    // TODO - LOG-2178: Add `isAvailable` for social plugins
+    // Use ["wechat-native", "alipay"] temporary
+    fun getInjectSocialScript() = """
+        window.logtoNativeSdk = {
+            platform: 'android',
+            getPostMessage: () => (data) => window.$NAME.postMessage(JSON.stringify(data)),
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:28]: image

File: android-sdk/android/src/main/AndroidManifest.xml (hunk 26 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/AndroidManifest.xml]: updated

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt (hunk 11 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/social/wechat/WechatSocialSession.kt]: Updated.
Remove this already, use "app_id" directly.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt (hunk 29 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt:8]: The linter finds that this is a global object, so it has a warning that this field may be leaked.
I think it's nothing to comment on here.

File: android-sample-kotlin/app/src/main/AndroidManifest.xml (hunk 23 above)
Comment (xiaoyijun) [reply android-sample-kotlin/app/src/main/AndroidManifest.xml:3]: 1. yes, this naming is commonly used in java and android. eg. `log4j`, `slf4j`
2. package name `kotlin` is only used for kotlin. our sample is good to place under the android package, for this sample is a sample of the android project which uses `android-sdk`, not `kotlin-sdk`.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt (hunk 25 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt]: refactored

File: android-sdk/android/src/main/AndroidManifest.xml (hunk 28 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/AndroidManifest.xml]: updated.

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt (hunk 30 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoAuthManager.kt]: updated. no need for this, for we capture this in the web-view

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt [hunk 32]
```diff
@@ -0,0 +1,64 @@
...
+            supportedSocialConnectorIds: ["wechat-native", "alipay"],
+            callbackUriScheme: '${hostActivity.packageName}.logto-callback-web',
```
Comment (gao-sun) [inline android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:30]: no need for this i think

**Review by gao-sun — APPROVED** [review]
overall lgtm

File: android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt (hunk 32 above)
Comment (xiaoyijun) [reply android-sdk/android/src/main/kotlin/io/logto/sdk/android/auth/logto/LogtoWebViewSocialHandler.kt:30]: LOG-2204

**Review by IceHe — APPROVED** [review]
lgtm
